dstep/translator/Context.d

```diff
@@ -198,9 +219,72 @@ bool shouldBeAnonymous(Context context, Cursor cursor)
 }
 
 /**
-  * Returns true, if cursor is in the global scope.
-  */
+ * Returns true, if cursor is in the global scope.
+ */
 bool isGlobal(Cursor cursor)
 {
     return cursor.semanticParent.kind == CXCursorKind.CXCursor_TranslationUnit;
 }
+
+/**
+ * The collectTypeNames function scans the whole AST of the translation unit and produces
+ * a set of the type names in global scope.
+ *
+ * The type names are required for parsing C code (e.g. macro definition bodies),
+ * as C grammar isn't context free.
+ */
+
+bool[string] collectTypeNames(TranslationUnit translUnit)
+{
+    void collectBasicTypes(ref bool[string] result)
+    {
+        result["void"] = true;
+        result["char"] = true;
+        result["signed char"] = true;
+        result["unsigned char"] = true;
+        result["short"] = true;
+        result["signed short"] = true;
+        result["unsigned short"] = true;
+        result["signed"] = true;
+        result["unsigned"] = true;
+        result["int"] = true;
+        result["signed int"] = true;
+        result["unsigned int"] = true;
+        result["long"] = true;
+        result["signed long"] = true;
+        result["unsigned long"] = true;
+        result["long long"] = true;
+        result["signed long long"] = true;
+        result["unsigned long long"] = true;
+        result["float"] = true;
+        result["double"] = true;
+    }
+
+    void collectTypeNames(ref bool[string] result, Cursor parent)
+    {
+        foreach (cursor, _; parent.all)
+        {
+            switch (cursor.kind)
+            {
+                case CXCursorKind.CXCursor_TypedefDecl:
+                    result[cursor.spelling] = true;
+                    break;
+
+                case CXCursorKind.CXCursor_StructDecl:
+                    result[cursor.spelling] = true;
+                    break;
+
+                default:
+                    break;
```

Comment:
Right, it is a bit leaky at the moment, there is probably a problem with tag namespace of strucutres (e.g. `struct Foo x;`). I'll fix it when I cope with constraining that `Args...` thing to be a string...